torchtune/modules/attention.py

```diff
@@ -227,3 +227,233 @@ def forward(
         # reshape the output to be the same shape as the input
         output = output.transpose(1, 2).contiguous().view(bsz, seq_len, -1)
         return self.output_proj(output)
+
+
+class GroupedQueryAttention(nn.Module):
```

Comment:
It's not clear to me why this is in modules/attention.py but other stuff goes in models/flamingo. I would suggest moving this under models/flamingo too to be consistent unless there's a clear reason not to.

Reply:
So our "primary" attention implementation already supports GQA and MQA [[ref](https://github.com/pytorch/torchtune/blob/main/torchtune/modules/attention.py#L189)]. I guess the question here is if this entire module is specific to flamingo, should this just reside within that folder till we figure out how to merge back? That said, I'll raise the same concern as I did in Philip's PR that GQA isn't a great name for this module. I'd prefer something like MMMultiHeadAttention or similar

Reply:
Yeah our naming here is kinda unclear. Like we already support GQA but via our `CausalSelfAttention` class. So any new functionality here is mainly to support cross-attention. So as is the naming of these two classes is not actually conveying how they differ. My two cents is that we should just call them `MultiHeadAttention` and `MultiHeadCrossAttention` to get this point across, but I think the renaming of `CausalSelfAttention` can be saved for another day.

Reply:
will rename it after we are 100% sure :P